controller/model/api_session_heartbeats.go

```diff
@@ -0,0 +1,143 @@
+/*
+	Copyright NetFoundry, Inc.
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package model
+
+import (
+	"github.com/michaelquigley/pfxlog"
+	"github.com/openziti/edge/controller/persistence"
+	"github.com/openziti/foundation/storage/boltz"
+	"github.com/openziti/foundation/util/concurrenz"
+	cmap "github.com/orcaman/concurrent-map"
+	"time"
+)
+
+type HeartbeatCollector struct {
+	apiSessionLastUpdate cmap.ConcurrentMap //map[apiSessionId string] => *heartbeatStatus
+	flushInterval        time.Duration
+	closeNotify          chan struct{}
+	isFlushing           concurrenz.AtomicBoolean
+	flushAction          func(apiSessionIds []string)
+	flushSize            int
+}
+
+// Creates a new HeartbeatCollector which is used to manage situations where an SDK is connecting to multiiple
+// Edge Routers and making API calls that all update their last updated at and trigger a write. The heartbeat
+// collector aggregates all of those calls into a single write and acts as an in memory buffer for last update
+// times.
+func NewHeartbeatCollector(env Env, flushSize int, flushInterval time.Duration, action func([]string)) *HeartbeatCollector {
+	collector := &HeartbeatCollector{
+		apiSessionLastUpdate: cmap.New(),
+		flushInterval:        flushInterval,
+		flushSize:            flushSize,
+		flushAction:          action,
+		closeNotify:          make(chan struct{}, 0),
+	}
+
+	env.GetStores().ApiSession.AddListener(boltz.EventDelete, collector.onDelete)
+
+	collector.Start()
+
+	return collector
+}
+
+type heartbeatStatus struct {
+	lastUpdatedAt time.Time
+	flushed       bool
+}
+
+func (self *HeartbeatCollector) Mark(apiSessionId string) {
+	newStatus := &heartbeatStatus{
+		lastUpdatedAt: time.Now(),
+		flushed:       false,
+	}
+	self.apiSessionLastUpdate.Set(apiSessionId, newStatus)
+}
+
+func (self *HeartbeatCollector) LastUpdatedAt(apiSessionId string) (time.Time, bool) {
+	if val, ok := self.apiSessionLastUpdate.Get(apiSessionId); ok {
+		if status, ok := val.(*heartbeatStatus); ok {
+			return status.lastUpdatedAt, true
+		}
+	}
+
+	return time.Time{}, false
+}
+
+func (self *HeartbeatCollector) Start() {
+	go self.run()
+}
+
+func (self *HeartbeatCollector) run() {
+	for {
+		select {
+		case <-self.closeNotify:
+			self.flush() //flush on stop
```

Comment:
On shutdown, flush. Graceful shutdowns need to work and hence the other shutdown work in this PR.